dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java

```diff
@@ -28,302 +28,224 @@
         {
-            jobConfiguration.setLastExecutedStatus( JobStatus.STOPPED );
-            jobConfigurationService.updateJobConfiguration( jobConfiguration );
-
-            internalStopJob( jobConfiguration.getUid() );
+            stoppedSuccessful( type );
         }
     }
 
     @Override
-    public boolean executeJob( JobConfiguration jobConfiguration )
+    public boolean executeNow( JobConfiguration configuration )
     {
-        if ( jobConfiguration != null && !isJobConfigurationRunning( jobConfiguration.getJobType() ) )
+        if ( configuration == null )
         {
-            internalExecuteJobConfiguration( jobConfiguration );
-            return true;
+            return false;
         }
-        else
+        JobType type = configuration.getJobType();
+        if ( type == null || isRunning( type ) )
         {
             return false;
         }
+        log.info( String.format( "Scheduler initiated execution of job: %s", configuration ) );
+        CompletableFuture<Future<?>> cancellation = new CompletableFuture<>();
+        Runnable task = runIfPossible( configuration, cancellation, () -> execute( configuration ) );
+        cancellation.complete( taskExecutor.executeTaskWithCancelation( task ) );
+        return true;
     }
 
-    @Override
-    public void executeJob( Runnable job )
-    {
-        jobExecutor.execute( job );
-    }
-
-    @Override
-    public Map<String, Future<?>> getAllFutureJobs()
-    {
-        return futures;
-    }
-
-    @Override
-    public Job getJob( JobType jobType )
-    {
-        return (Job) applicationContext.getBean( jobType.getKey() );
-    }
-
-    @Override
-    public <T> ListenableFuture<T> executeJob( Callable<T> callable )
-    {
-        return jobExecutor.submitListenable( callable );
-    }
-
-    // -------------------------------------------------------------------------
-    // Supportive methods
-    // -------------------------------------------------------------------------
-
-    private void internalExecuteJobConfiguration( JobConfiguration jobConfiguration )
-    {
-        JobInstance jobInstance = new DefaultJobInstance( this, messageService, leaderManager );
-
-        ListenableFuture<?> future = jobExecutor.submitListenable( () -> jobInstance.execute( jobConfiguration ) );
-
-        String uid = jobConfiguration.getUid();
-        currentTasks.put( uid, future );
-        future.completable().whenComplete( ( res, ex ) -> currentTasks.remove( uid ) );
-
-        log.info( String.format( "Scheduler initiated execution of job: %s", jobConfiguration ) );
-    }
-
-    private boolean internalStopJob( String uid )
+    /**
+     * Wraps the original task in order to manage cancellation state correctly
+     * as part of running the task so that transitions occur when the task
+     * actually executes.
+     */
+    private Runnable runIfPossible( JobConfiguration configuration, Future<Future<?>> cancellation, Runnable task )
```

Comment:
The `cancellation` parameter here surely is a bit mind bending because we have a hen-egg situation.
The `task` we want to execute should be aware of `Future` we get to cancel its execution so it can safely remove itself from the internal data structures but the `Future` needed is first returned by the executor when we run/schedule the task. To overcome this we pass in a promise of this cancellation `Future` in form of a `Future` which then then `complete` as soon as it has been returned by the executer used.